### PR TITLE
Keystore overhaul

### PR DESCRIPTION
**Pull request by davxy**
Partially address https://github.com/paritytech/substrate/issues/13556

This PR is part of a small overhaul of the keystore related traits and structures.

A followup PR will focus on the keystore methods... I prefer to keep this PR small and quickly reviewable

---

Mostly delivers:
- removal of the `CryptoStore`async trait. 
  It was introduced a long ago in the hope of a remote keystore.
  Today the 90% of our code still depend on the sync version thus there's not point of having it around. If *one day™️* it will be really required then well can trivially reintroduce it. 
  But in the meantime is better to have no bloat in an already giant repo.
- trivial renaming to better express the traits intent in a less verbose way. May be a bit opinionated but I think that overall the introduced names are a bit more coherent:
  - trait `SyncCryptoStore` -> `Keystore`
  - trait `CryptoStore` → removed 🔴
  - `SyncCryptoKeyStorePtr` → `KeystorePtr`
  - struct `testing::KeyStore` → `testing::MemoryKeystore` (this is what it is and is more inline with the other implementation we provide: `LocalKeystore`).
- removal of a couple of unused methods from the trait interface: `sign_with_any`  and `sign_with_all`
   - the former was unused
   - the latter was used in one single place and can be trivially performed in the iteration that was required anyway for additional operations in the caller.

---

NOTE: big part of the changes spread through the project are due to the renamings 🫤. 

polkadot companion: https://github.com/paritytech/polkadot/pull/6892
cumulus companion: https://github.com/paritytech/cumulus/pull/2336

### CODE REVIEW

**Review by melekes — APPROVED** [review]
👍

File: primitives/keystore/src/lib.rs [hunk 1]
```diff
@@ -257,6 +84,7 @@ pub trait SyncCryptoStore: CryptoStore + Send + Sync {
 	fn ecdsa_generate_new(&self, id: KeyTypeId, seed: Option<&str>)
 		-> Result<ecdsa::Public, Error>;
 
+	/// TODO-DAVXY: THIS should be renamed as "insert"... a key crypto type is never known
```
Comment (skunert) [inline primitives/keystore/src/lib.rs:87]: I think you could include it in this PR

File: client/service/src/builder.rs [hunk 2]
```diff
@@ -86,18 +86,19 @@ type TFullParts<TBl, TRtApi, TExec> =
 	(TFullClient<TBl, TRtApi, TExec>, Arc<TFullBackend<TBl>>, KeystoreContainer, TaskManager);
 
 trait AsCryptoStoreRef {
-	fn keystore_ref(&self) -> Arc<dyn CryptoStore>;
-	fn sync_keystore_ref(&self) -> Arc<dyn SyncCryptoStore>;
+	// TODO-DAVXY: this can be removed
```
Comment (skunert) [inline client/service/src/builder.rs:89]: Leftover todo? Could be included here?

File: primitives/keystore/src/lib.rs (hunk 1 above)
Comment (davxy) [reply primitives/keystore/src/lib.rs]: done

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2533958

**Review by skunert — APPROVED** [review]

**Comment by davxy** [comment]
Polkadot CI fails because of https://github.com/paritytech/polkadot/pull/6898

**Comment by davxy** [comment]
bot merge

**Review by michalkucharczyk — APPROVED** [review]

**Comment by davxy** [comment]
bot merge

**Pull request merged**